File: mayonez-base/src/main/java/mayonez/graphics/renderer/DrawPrimitive.java
package mayonez.graphics.renderer;

import mayonez.annotations.EngineType;
import mayonez.annotations.UsesEngine;
import mayonez.math.IntMath;
import org.lwjgl.opengl.GL11;

import java.nio.IntBuffer;

/**
 * Types of primitive objects submitted to the GPU. Only supported by GL.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
public enum DrawPrimitive {
    /**
     * An object with 2 vertices with attributes Pos (2), Color (4)
     */
    LINE(2, 2, GL11.GL_LINES, 2, 4) {
        @Override
        public void addIndices(IntBuffer elements, int index) {
            for (int v = 0; v <= 1; v++) elements.put(vertexCount * index + v);
        }
    },
    /**
     * An object with 3 vertices with attributes Pos (2), Color (4)
     */
    TRIANGLE(3, 3, GL11.GL_TRIANGLES, 2, 4) {
        @Override
        public void addIndices(IntBuffer elements, int index) {
            // Use counterclockwise winding
            for (int v = 0; v <= 2; v++) elements.put(vertexCount * index + v);
        }
    },
    /**
     * An object with 4 vertices with attributes Pos (2), Color (4), Tex Coords (2), Tex ID (1)
     */
    SPRITE(4, 6, GL11.GL_TRIANGLES, 2, 4, 2, 1) {
        @Override
        public void addIndices(IntBuffer elements, int index) {
            // Split quad into two triangles
            int[] triangleVertices = {0, 1, 2, 0, 2, 3};
            for (int v : triangleVertices) elements.put(vertexCount * index + v);
        }
    };

    // Vertex Parameters
    final int vertexCount; // Vertices defined by primitive
    final int elementCount; // Elements needed by GL (i.e. split quad into 2 triangles)
    final int primitiveType; // Primitive ID used by OpenGL
    final int[] attributeSizes; // floats inside each attribute
    final int vertexSize; // Floats inside one vertex

    DrawPrimitive(int vertexCount, int elementCount, int primitiveType, int... attributeSizes) {
        this.vertexCount = vertexCount;
        this.elementCount = elementCount;
        this.primitiveType = primitiveType;
        this.attributeSizes = attributeSizes;
        vertexSize = IntMath.sum(attributeSizes);
    }

    /**
     * Adds indices to an element buffer array (EBO).
     *
     * @param elements an int buffer
     * @param index    the vertex index
     */
    public abstract void addIndices(IntBuffer elements, int index);

}

File: mayonez-base/src/main/kotlin/mayonez/renderer/batch/DrawPrimitive.kt
package mayonez.renderer.batch

import mayonez.graphics.*
import java.nio.IntBuffer

// Vertex Attribute Aliases
private val POSITION: VertexAttribute = VertexAttribute.FLOAT2 // (x, y, 0)
private val TEX_COORD: VertexAttribute = VertexAttribute.FLOAT2 // (u, v)
private val COLOR: VertexAttribute = VertexAttribute.FLOAT4 // (r, g, b, a)
private val TEX_SLOT: VertexAttribute = VertexAttribute.INT // i

/**
 * Types of OpenGL primitive objects that can be submitted to the GPU. Each
 * object defines the element and attribute layout.
 *
 * @author SlavSquatSuperstar
 */
@UsesEngine(EngineType.GL)
enum class DrawPrimitive(
    val layout: ElementLayout, vararg val attributes: VertexAttribute
) {

    /**
     * A triangle with 3 vertices, each with attributes position and color.
     */
    TRIANGLE(
        ElementLayout.TRIANGLE,
        POSITION, COLOR
    ),

    /**
     * An untextured textured quadrangle with 4 vertices, each with attributes
     * position and color.
     *
     * Note that lines are drawn using quads, since glLineWidth() does not support
     * line widths above 1 on all platforms.
     *
     * Sources:
     * - [Drawing lines with shaders](https://stackoverflow.com/questions/3484260/opengl-line-width)
     * - [Drawing lines as triangles](https://stackoverflow.com/questions/101718/drawing-a-variable-width-line-in-opengl-no-gllinewidth)
     * - [Drawing OpenGL lines](https://stackoverflow.com/questions/14486291/how-to-draw-line-in-opengl)
     */
    QUAD(
        ElementLayout.QUAD,
        POSITION, COLOR
    ),

    /**
     * A textured quadrangle with 4 vertices, each with attributes position, color,
     * texture coordinate, and texture slot.
     */
    SPRITE(
        ElementLayout.QUAD,
        POSITION, COLOR, TEX_COORD, TEX_SLOT
    ),

    /**
     * A circle with 4 vertices, each with attributes position, local position,
     * color, and inner radius.
     *
     * Sources:
     * - [Drawing circles as triangles](https://www.youtube.com/watch?v=VEnglRKNHjU)
     * - [Drawing circles as quads](https://stackoverflow.com/questions/22444450/drawing-circle-with-opengl/50408198#50408198)
     * - [TheCherno Hazel Engine circle shader](https://github.com/TheCherno/Hazel/blob/master/Hazelnut/assets/shaders/Renderer2D_Circle.glsl)
     */
    CIRCLE(
        ElementLayout.QUAD,
        POSITION, POSITION, COLOR, VertexAttribute.FLOAT
    ),

    /**
     * An ellipse with 4 vertices, each with attributes position, local position,
     * color, and two inner radii.
     */
    ELLIPSE(
        ElementLayout.QUAD,
        POSITION, POSITION, COLOR, VertexAttribute.FLOAT2
    );

    val vertexCount: Int
        get() = layout.vertexCount

    val elementCount: Int
        get() = layout.elementCount

    val drawMode: Int
        get() = layout.drawMode

    /** The number of components for all attributes per vertex. */
    val totalComponents: Int = attributes.sumOf { it.components }

    /**
     * Adds indices to an element buffer array (EBO).
     *
     * @param elements an int buffer
     * @param index the vertex index
     */
    fun addIndices(elements: IntBuffer, index: Int) {
        for (e in layout.elementOrder) {
            elements.put(vertexCount * index + e)
        }
    }

}